File: xamarin/common/Presenter/Services/PropertyFinderPersistentState.cs
using System.Collections.Generic;
using PropertyFinder.Model;
using System.Linq;
using System.Xml.Serialization;
using System.IO;
using System.Diagnostics;
using System.Runtime.Serialization;

namespace PropertyFinder.Presenter
{
  /// <summary>
  /// This class hold all the application state that should be persisted
  /// between sessions. When the state changes, this class persists itself via
  /// the supplied <see cref="IStatePersistenceService"/> instance.
  /// </summary>
  [XmlInclude(typeof(PlainTextSearchItem))]
  [XmlInclude(typeof(GeoLocationSearchItem))]
  public class PropertyFinderPersistentState
  {
    private IStatePersistenceService _persistenceService;

    public IStatePersistenceService PersistenceService
    {
      set { _persistenceService = value; }
    }

    public PropertyFinderPersistentState()
    {
      Favourites = new List<Property>();
      RecentSearches = new List<RecentSearch>();
    }

    public PropertyFinderPersistentState(IStatePersistenceService persistenceService)
      : this()
    {
      _persistenceService = persistenceService;
    }

    /// <summary>
    /// Gets or sets the properties the user has marked as favourites
    /// </summary>
    public List<Property> Favourites { get; set; }

    /// <summary>
    /// Gets or sets a few recent searches
    /// </summary>
    public List<RecentSearch> RecentSearches { get; set; }

    /// <summary>
    /// Adds the given search to the RecentSearches list. 
    /// </summary>
    public void AddSearchToRecent (RecentSearch search)
    {
      // if we already have this search saved, move it to the top
      if (RecentSearches.Any (r => r.Search.DisplayText == search.Search.DisplayText))
      {
        var matchingSearch = RecentSearches.Single (r => r.Search.DisplayText == search.Search.DisplayText);
        RecentSearches.Remove (matchingSearch);
        RecentSearches.Insert (0, matchingSearch);
      }
      else
      {
        // otherwise, add it
        RecentSearches.Insert (0, search);
        if (RecentSearches.Count > 4)
        {
          RecentSearches.RemoveAt (RecentSearches.Count - 1);
        }
      }
      _persistenceService.SaveState(this);
    }

    /// <summary>
    /// Gets whether the given property is favourited
    /// </summary>
    public bool IsPropertyFavourited(Property property)
    {
      return Favourites.Any(p => p.Guid == property.Guid);
    }

    /// <summary>
    /// Toggles the favourited state of the given property.
    /// </summary>
    public void ToggleFavourite(Property property)
    {
      if (IsPropertyFavourited(property))
      {
        var matchingProperty = Favourites.Single(p => p.Guid == property.Guid);
        Favourites.Remove(matchingProperty);
      }
      else
      {
        Favourites.Add(property);
      }

      PersistState();
    }

    /// <summary>
    /// Persists the current state.
    /// </summary>
    private void PersistState()
    {
      _persistenceService.SaveState(this);
    }
  }
}

File: native/android/PropertyCross/src/com/propertycross/android/presenter/PropertyFinderPersistentState.java
package com.propertycross.android.presenter;

import java.util.ArrayList;
import java.util.List;

import com.propertycross.android.model.Property;

public class PropertyFinderPersistentState {

	private IStatePersistenceService service;
	private List<Property> favourites;
	private List<RecentSearch> recentSearches;
	
	public PropertyFinderPersistentState() {
		favourites = new ArrayList<Property>();
		recentSearches = new ArrayList<RecentSearch>();
	}
	
	public PropertyFinderPersistentState(IStatePersistenceService value) {
		favourites = new ArrayList<Property>();
		recentSearches = new ArrayList<RecentSearch>();
		service = value;
	}
	
	public void setPersistenceService(IStatePersistenceService value) {
		service = value;
	}
	
	public List<Property> getFavourites() {
		return favourites;
	}
	
	public void setFavourites(List<Property> value) {
		favourites = value;
	}
	
	public List<RecentSearch> getRecentSearches() {
	    return recentSearches;
	}
	
	public void setRecentSearches(List<RecentSearch> value) {
		recentSearches = value;
	}
	
	public void addSearchToRecent(RecentSearch search) {
		RecentSearch matchingSearch = getMatchingRecentSearchFor(search);
		if (matchingSearch != null) {
			recentSearches.remove(matchingSearch);
			((ArrayList<RecentSearch>)recentSearches).add(0, matchingSearch);
		}
		else {
			((ArrayList<RecentSearch>) recentSearches).add(0, search);
			if (recentSearches.size() > 4) {
				recentSearches.remove(recentSearches.size() - 1);
			}
		}
	}
	
	private RecentSearch getMatchingRecentSearchFor (RecentSearch s) {
		for (RecentSearch search : recentSearches) {
			if (search.getSearch().getDisplayText().equals(s.getSearch().getDisplayText())) {
				return search;
			}
		}
		return null;
	}
	
	public boolean isPropertyFavourited(Property p) {
		for(Property property : favourites) {
			if(p.getGuid() == property.getGuid()) {
				return true;
			}
		}		
		return false;
	}
	
	public void toggleFavourite(Property p) {
		
		if (isPropertyFavourited(p)) {
			favourites.remove(getMatchingPropertyFor(p));
		}
		else {
			favourites.add(p);
		}
		
		persistState();		
	}
	
	private Property getMatchingPropertyFor(Property p) {
		for (Property property : favourites) {
			if (p.getGuid() == property.getGuid()) {
				return property;
			}
		}
		
		return null;
	}
	
	private void persistState() {
		service.saveState(this);
	}
}